### PR TITLE
[ACR] Add validations for blob and manifest download

### LINKED ISSUE #35771: Add validation to the download blob API.
Given the download blob can redirect. We want to add the following validation to the download blob API

- [x] Do not validate docker-content-digest header for blob download (but keep validating requested digest vs content)
- [x] Do validate content-range header when downloading blob

> [ ] it must be present and have positive full length. throw exception otherwise
> [ ] only read up to this length

- [x] Validate manifest when downloading it:

> - [x] check docker-content-digest header (even if manifest is requested by digest)
> - [x] may add sanity check for manifest size - 4mb is unofficial, but justified [containerd] (https://github.com/containerd/containerd/blob/main/remotes/docker/resolver.go#L59) limit

Java PR - https://github.com/Azure/azure-sdk-for-java/pull/34737

### PR DESCRIPTION
**Pull request by annelo-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/35771

### CODE REVIEW

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs [hunk 1]
```diff
@@ -671,6 +680,28 @@ private static bool ReferenceIsDigest(string reference)
             return reference.StartsWith("sha256:", StringComparison.OrdinalIgnoreCase);
         }
 
+        private static void CheckBlobSize(Response response)
```
Comment (lmolkova) [inline sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs:683]: nit (to reuse it in manifest validation):
```suggestion
        private static void CheckContentLength(Response response)
```

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs [hunk 2]
```diff
@@ -671,6 +680,28 @@ private static bool ReferenceIsDigest(string reference)
...
+                throw new RequestFailedException(response.Status, InvalidContentLengthMessage);
+            }
+        }
+
+        private static void CheckManifestSize(Response response)
+        {
+            // This check is to address part of the service threat model.
+            // If a manifest does not have a proper content length or is too big,
+            // it indicates a malicious or faulty service and should not be trusted.
+            int? size = response.Headers.ContentLength ?? throw new RequestFailedException(response.Status, InvalidContentLengthMessage);
```
Comment (lmolkova) [inline sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs:697]: nit:
```suggestion
            CheckContentLength(response);
            int? size = response.Headers.ContentLength;
```

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs [hunk 3]
```diff
@@ -532,10 +536,27 @@ private static string GetContentRange(long offset, long length)
...
         {
-            string size = contentRange.Split('/')[1];
-            return long.Parse(size, CultureInfo.InvariantCulture);
+            if (!response.Headers.TryGetValue("Content-Range", out string contentRange) ||
+                contentRange == null)
+            {
+                throw new RequestFailedException(response.Status, InvalidContentRangeMessage);
+            }
+
+            string[] values = contentRange.Split('/');
```
Comment (lmolkova) [inline sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs:547]: nit: it's a bit cheaper to do `IndexOf('\')` and substring the seconds part

**Review by lmolkova — APPROVED** [review]
LGTM! just left a few minor comments

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Content/ContainerRegistryContentClient.cs:547]: TY!

**Pull request merged**